src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.EnumerateSplits.cs

```diff
@@ -0,0 +1,270 @@
+        /// <see cref="Split(string)"/> will include the contents of capture groups in the resulting splits, while <see cref="EnumerateSplits(ReadOnlySpan{char})"/> will not.
+        /// And if <see cref="RegexOptions.RightToLeft"/> is specified, <see cref="Split(string)"/> will reverse the order of the resulting splits to be left-to-right, whereas
+        /// <see cref="EnumerateSplits(ReadOnlySpan{char})"/> will yield the splits in the order they're found right-to-left.
+        /// </para>
+        /// <para>
+        /// Each match won't actually happen until <see cref="ValueSplitEnumerator.MoveNext"/> is invoked on the enumerator, with one match being performed per <see cref="ValueSplitEnumerator.MoveNext"/> call.
+        /// Since the evaluation of the match happens lazily, any changes to the passed in input in between calls to <see cref="ValueSplitEnumerator.MoveNext"/> will affect the match results.
+        /// </para>
+        /// </remarks>
+        /// <param name="input">The span to search for a match.</param>
+        /// <returns>A <see cref="ValueSplitEnumerator"/> to iterate over the matches.</returns>
+        public ValueSplitEnumerator EnumerateSplits(ReadOnlySpan<char> input) =>
+            EnumerateSplits(input, count: 0);
+
+        /// <summary>
+        /// Searches an input span for all occurrences of a regular expression and returns a <see cref="ValueSplitEnumerator"/> to iterate over the splits around matches.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The behavior of <see cref="EnumerateSplits(ReadOnlySpan{char}, int)"/> is similar to the behavior of <see cref="Split(string, int)"/>, producing the splits
+        /// one at a time as part of iterating through the resulting enumerator rather than all at once as part of a single array. However, there are a few notable differences.
+        /// <see cref="Split(string, int)"/> will include the contents of capture groups in the resulting splits, while <see cref="EnumerateSplits(ReadOnlySpan{char}, int)"/> will not.
+        /// And if <see cref="RegexOptions.RightToLeft"/> is specified, <see cref="Split(string, int)"/> will reverse the order of the resulting splits to be left-to-right, whereas
+        /// <see cref="EnumerateSplits(ReadOnlySpan{char}, int)"/> will yield the splits in the order they're found right-to-left.
+        /// </para>
+        /// <para>
+        /// Each match won't actually happen until <see cref="ValueSplitEnumerator.MoveNext"/> is invoked on the enumerator, with one match being performed per <see cref="ValueSplitEnumerator.MoveNext"/> call.
+        /// Since the evaluation of the match happens lazily, any changes to the passed in input in between calls to <see cref="ValueSplitEnumerator.MoveNext"/> will affect the match results.
```

Comment:
affect the match results potentially causing an invalid result, right? 

Although this is lazy, one could imagine some future optimization someday (some kind of pre scanning or memoization or whatnot, who knows) that made it less than lazy in effect, such a change to the input span even if it was "ahead of" the current match would cause invalid results..? Should we document that it's not supported to do this?

Reply:
> affect the match results potentially causing an invalid result, right?

Depends on how you define invalid. It just starts a new match from where it left off, so anything changed between MoveNexts will result in it just processing the new data, whatever that means.

> Should we document that it's not supported to do this?

Ok.

Reply:
> Depends on how you define invalid. It just starts a new match from where it left off, so anything changed between MoveNexts will result in it just processing the new data, whatever that means.

That's true today, but I was hypothesizing that we might want to make performance optimizations in future, perhaps years from now, that would break that - some kind of precalculation. Assuming the 95% case is not mutating the text during iteration, documenting that we don't support mutation allows us to defer the decision of whether we want to be constrained that way.

Reply:
Which is why I said ok :)